extension/src/time_weighted_average.rs

```diff
@@ -0,0 +1,422 @@
+pub fn time_weight_trans(
+    state: Option<Internal<TimeWeightTransState>>,
+    method: String,
+    ts: Option<pg_sys::TimestampTz>,
+    val: Option<f64>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal<TimeWeightTransState>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let p = match (ts, val) {
+                (_, None) => return state,
+                (None, _) => return state, 
+                (Some(ts), Some(val)) => TSPoint{ts, val},
+            };
+            let method = match method.to_lowercase().as_str() {
+                "linear"=> TimeWeightMethod::Linear,
+                "locf" => TimeWeightMethod::LOCF,
+                &_ => panic!("unknown method")
+            };
+            match state {
+                None => {let mut s = TimeWeightTransState{buffer: vec![], method: None, combine_buffer: vec![] }; s.push(p, method); Some(s.into())},
+                Some(mut s) => {s.push(p, method); Some(s)},
+            }
+        })
+    }
+}
+
+
+#[pg_extern(schema = "timescale_analytics_experimental")]
+pub fn time_weight_summary_trans(
+    state: Option<Internal<TimeWeightTransState>>,
+    next: Option<time_weight_summary>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal<TimeWeightTransState>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state, next) {
+                (None, None) => None,
+                (None, Some(next)) => Some(TimeWeightTransState{combine_buffer:vec![next.to_TimeWeightSummary()], buffer: vec![], method: Some(*next.method)}.into()),
+                (Some(state), None) => Some(state),
+                (Some(mut state), Some(next)) =>  {
+                    let next = TimeWeightTransState{combine_buffer:vec![next.to_TimeWeightSummary()], buffer: vec![], method: Some(*next.method)};
+                    state.combine_push(&next); 
+                    Some(state.into())
+                },
+            }
+        })
+    }
+}
+
+
+#[pg_extern(schema = "timescale_analytics_experimental")]
+pub fn time_weight_combine(
+    state1: Option<Internal<TimeWeightTransState>>,
+    state2: Option<Internal<TimeWeightTransState>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+)  -> Option<Internal<TimeWeightTransState>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state1, state2) {
+                (None, None) => None,
+                (None, Some(state2)) => {let mut s = state2.clone(); s.sort_and_calc(); Some(s.into())},
+                (Some(state1), None) => {let mut s = state1.clone(); s.sort_and_calc(); Some(s.into())}, //should I make these return themselves?
+                (Some(state1), Some(state2)) => {
+                    let mut s1 = state1.clone(); // is there a way to avoid if it doesn't need it?
+                    s1.sort_and_calc(); 
+                    let mut s2 = state2.clone();
+                    s2.sort_and_calc();
+                    s2.combine_push(&s1);
```

Comment:
Should we ensure these have the same TimeWeightMethod here, or inside this function?  I think right now it won't complain until someone calls `combine_sort_and_calc` at some point in the future.

Reply:
Yeah. checking this is a bit weird right now. I think I want to modify when that happens in general, I'm going to leave that as outstanding for now.